src/Platform.Handlers/src/Xamarin.Platform.Handlers/Hosting/Internal/AppHost.cs

```diff
@@ -0,0 +1,147 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Hosting.Internal;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Xamarin.Platform.Hosting.Internal
+{
+	internal class AppHost : IHost, IAsyncDisposable
+	{
+		readonly ILogger<AppHost> _logger;
+		readonly IHostLifetime _hostLifetime;
+		readonly ApplicationLifetime _applicationLifetime;
+		readonly HostOptions _options;
+		IEnumerable<IHostedService>? _hostedServices;
+
+		public AppHost(IServiceProvider services, IHostApplicationLifetime applicationLifetime, ILogger<AppHost> logger,
+			IHostLifetime hostLifetime, IOptions<HostOptions> options)
+		{
+			Services = services ?? throw new ArgumentNullException(nameof(services));
+			_applicationLifetime = (applicationLifetime as ApplicationLifetime) ?? throw new ArgumentNullException(nameof(applicationLifetime));
+			if (_applicationLifetime is null)
+			{
+				throw new ArgumentException("Replacing IHostApplicationLifetime is not supported.", nameof(applicationLifetime));
+			}
+			_logger = logger ?? throw new ArgumentNullException(nameof(logger));
+			_hostLifetime = hostLifetime ?? throw new ArgumentNullException(nameof(hostLifetime));
+			_options = options?.Value ?? throw new ArgumentNullException(nameof(options));
+		}
+
+		public IServiceProvider Services { get; }
+
+		public async Task StartAsync(CancellationToken cancellationToken = default)
+		{
+			_logger.Starting();
+
+			using var combinedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _applicationLifetime.ApplicationStopping);
+			CancellationToken combinedCancellationToken = combinedCancellationTokenSource.Token;
+
+			await _hostLifetime.WaitForStartAsync(combinedCancellationToken).ConfigureAwait(false);
+
+			combinedCancellationToken.ThrowIfCancellationRequested();
+			_hostedServices = Services?.GetService<IEnumerable<IHostedService>>();
+			if (_hostedServices != null)
+			{
+				foreach (IHostedService hostedService in _hostedServices)
+				{
+					// Fire IHostedService.Start
+					await hostedService.StartAsync(combinedCancellationToken).ConfigureAwait(false);
+
+					if (hostedService is BackgroundService backgroundService)
+					{
+						//_ = HandleBackgroundException(backgroundService);
```

Comment:
The reason we do this in Hosting is because exceptions that occur after an `await` in a BackgroundService aren't handled at all. So in Hosting, we handle them here and write it to the log.

In a mobile application, maybe it should crash the app?